### DISCUSSION TITLE
How to show Bindings from Tabs in TabbedContent

### DISCUSSION THREAD

**Discussion by jakubziebin** [Textualize/textual#3419] [Q&A]
I am using a TabPane with TabbedContent. I want to show Bindings from Tabs, so `Binding("left", "previous_tab", "Previous tab", show=False), Binding("right", "next_tab", "Next tab", show=False)`, but only when I focus on Tabs. Is there an elegant way to do this without going into too much detail?

**Comment by davep** [comment]
Unfortunately I don't think there's an easy approach to this at the moment.

**Reply by jakubziebin** [reply]
Hm, so something  more in depth ? I thought about using `query_one` `Tabs` and set `Bindings` that belong to them to `show=True`, but i do not have any idea how to catch focus that is on `Tabs`, and unfortunately `query_one` does not catch `Tabs`.

**Reply by TomJGooding** [reply]
There's no way of changing bindings while the app is running, AFAIK.

Generally you would inherit from a widget and add your own bindings. Here's an example from discussion #3256 to show the `ctrl+u` binding for an `Input` widget:

```python
class MyInput(Input):
    BINDINGS = [
        Binding("ctrl+u", "delete_left_all", "Delete all to the left", show=True),
    ]
```

The difficulty is that `TabbedContent` is a [compound widget](https://textual.textualize.io/guide/widgets/#compound-widgets) composed of `Tabs` and `ContentSwitcher`. You could create a custom `Tabs` class with the modified bindings. but obviously then you would need a custom `TabbedContent` widget instead composed with your custom Tabs.

**Reply by mzebrak** [reply]
> There's no way of changing bindings while the app is running, AFAIK.

This is a feature that would help in many situations like this. I've also run into  issues that lack of dynamic bindings led me doing weird things: https://github.com/Textualize/textual/issues/3041#issuecomment-1660080726. 

What I found is this issue https://github.com/Textualize/textual/issues/1057, maybe this discussion is yet another reason to have such a dynamic bindings?

**Reply by TomJGooding** [reply]
@mzebrak I'm not a maintainer of Textual, but perhaps rather than just posting on anything than mentions bindings, it would be more useful if you could suggest a possible implementation or API? Especially because 'dynamic' bindings don't necessarily mean the same thing across all the threads you've created notifications for.

**Reply by mzebrak** [reply]
> @mzebrak I'm not a maintainer of Textual, but perhaps rather than just posting on anything than mentions bindings, it would be more useful if you could suggest a possible implementation or API? Especially because 'dynamic' bindings don't necessarily mean the same thing across all the threads you've created notifications for.

Sorry for the late response, I think I missed that and landed back there again today when looking at the https://textual.textualize.io/guide/actions/#dynamic-actions feature which does not allow for adding/removing bindings in a dynamic way. From my understanding, it just allows for blocking some existing binding or [updating its description](https://github.com/Textualize/textual/issues/4755).

What we currently need and have implemented (unfortunately using Textual internals) is basically:
```python
    def bind(self, binding: Binding) -> None:
        self._bindings.key_to_bindings.setdefault(binding.key, []).append(binding)
        self.refresh_bindings()

    def unbind(self, key: str) -> None:
        binding = self._bindings.key_to_bindings.pop(key, None)
        if binding:
            self.refresh_bindings()

```

at the Widget level. This allows us to bind also at the screen because Screen inherits from Widget.

To be completely happy, I would like to see something like this in Textual and to be able to manage bindings at the level of the  application (global bindings), screen, and widget.

I think this will suit all the possible use-cases and there is no need to differentiate 
>  Especially because 'dynamic' bindings don't necessarily mean the same thing

You can just remove any existing binding and add any new binding. If you want to update the description, you can unbind and bind again or maybe just bind to the same key.

Edit
---

There is already a way to dynamically add a global binding:
```python
from __future__ import annotations

from textual.app import App, ComposeResult
from textual.binding import Binding
from textual.widgets import Footer, Label


class MyApp(App):
    BINDINGS = [
        Binding("a", "add_binding", "Add binding"),
    ]

    def compose(self) -> ComposeResult:
        yield Footer()

    def action_add_binding(self) -> None:
        self.bind("m", "mount_label", description="Mount label")
        self.refresh_bindings()

    def action_mount_label(self) -> None:
        self.mount(Label("Hello, World!"), before="Footer")


MyApp().run()
```

But it's impossible or I don't see a way to remove a binding or to do it more locally (Screen, Widget).


Edit2
--- 

> perhaps rather than just posting on anything than mentions bindings, it would be more useful if you could suggest a possible implementation or API?

[I've actually proposed an interface](https://github.com/Textualize/textual/discussions/1792#discussioncomment-5265859 ) before coming back to this discussion.  

Now I'm thinking that just the minimal `unbind`/`bind` at `App`/`Screen`/`Widget` level is sufficient.
